Update release notes for oneTBB 2021.7 (#872)
oneapi-src/oneTBB: RELEASE_NOTES.md

This document contains changes of oneTBB compared to the last release.

## Table of Contents <!-- omit in toc -->
- [Known Limitations](#known-limitations)
- [Fixed Issues](#fixed-issues)
- [Open-source Contributions Integrated](#open-source-contributions-integrated)

## :rotating_light: Known Limitations

> - A static assert will cause compilation failures in oneTBB headers when compiling with clang 12.0.0 or newer if using the LLVM standard library with -ffreestanding and C++11/14 compiler options.

Reviewer: ```suggestion
- A static assert causes compilation failures in oneTBB headers when compiling with Clang* 12.0.0 or newer. The reason is using the LLVM* standard library with -ffreestanding and C++11/14 compiler options.
```

Reviewer 2: @alexandraepan the compilation failure is only seen if LLVM 12 or newer is used and the -ffreestanding option is used along with -std=c++11 or -std=c++14. The proposed wording reads as though the assert fails in all cases where Clang 12.0.0 or newer is used.

Reviewer: @timmiesmith okay, thanks for the clarification

Reviewer: Updated the comment

- C++ exception handling mechanism on Windows* OS on ARM64* might corrupt memory if an exception is thrown from any oneTBB parallel algorithm (see Windows* OS on ARM64* compiler issue: https://developercommunity.visualstudio.com/t/ARM64-incorrect-stack-unwinding-for-alig/1544293).

## :hammer: Fixed Issues
- Memory allocator crash when allocating ~1TB on 64-bit systems (GitHub* [#838](https://github.com/oneapi-src/oneTBB/issues/838)).
- Thread Distribution over NUMA Nodes on Windows OS systems.

> - The function provided to oneapi::tbb::suspend is now guaranteed to be executed on the same thread.

Reviewer 3: ```suggestion
- The user function provided to oneapi::tbb::suspend is now guaranteed to be called on the same thread that called oneapi::tbb::suspend.
```

Reviewer 4: May be it would be even better if we remove "`the same`":
`The user function provided to oneapi::tbb::suspend function is now guaranteed to be called on the thread that called oneapi::tbb::suspend`
What do you think?

Reviewer 5: Maybe "For oneapi::tbb::suspend it is now guaranteed that the user-specified callable object is executed by the calling thread"
@pavelkumbrasev Is there a specification update patch that matches this change, or do we consider it an implementation specific behavior?

## :octocat: Open-source Contributions Integrated

> - Fix for full LTO build (library and tests) on UNIX* OS systems. Contributed by Vladislav Shchapov (https://github.com/oneapi-src/oneTBB/pull/671).

Reviewer 6: PR: #798

- C++ exception handling mechanism on Windows* OS on ARM64* might corrupt memory if an exception is thrown from any oneTBB parallel algorithm (see Windows* OS on ARM64* compiler issue: https://developercommunity.visualstudio.com/t/ARM64-incorrect-stack-unwinding-for-alig/1544293).

## :hammer: Fixed Issues
- Memory allocator crash when allocating ~1TB on 64-bit systems (GitHub* [#838](https://github.com/oneapi-src/oneTBB/issues/838)).

> - Thread Distribution over NUMA Nodes on Windows OS systems.

Author: @akukanov, @alexandraepan, I'm not sure if we need word "Fixed" in the beginning. It looks the section format and naming make it redundant, at least it's not a part of other notes in the same section.

> - Thread Distribution over NUMA Nodes on Windows OS systems.

Reviewer 5: If not the word "fixed", some other rewording is needed. In my opinion, any note should be self-descriptive, you do not have to know what section it is in to understand it.

> - Thread Distribution over NUMA Nodes on Windows OS systems.

Author: I've applied Alexey's suggestion

> - For oneapi::tbb::suspend it is now guaranteed that the user-specified callable object is executed by the calling thread.

Reviewer: ```suggestion
- For oneapi::tbb::suspend, it is guaranteed that the user-specified callable object is executed by the calling thread.
```

## :octocat: Open-source Contributions Integrated

> - Fix for full LTO build (library and tests) on UNIX* OS systems. Contributed by Vladislav Shchapov (https://github.com/oneapi-src/oneTBB/pull/798).

Reviewer: ```suggestion
- Fix for full LTO* build, library and tests, on UNIX* OS systems. Contributed by Vladislav Shchapov (https://github.com/oneapi-src/oneTBB/pull/798).
```
